File: WebClient SDK/src/default/ui/krProcess/stageHandlers/processManagementUIHandler.ts
import { DefaultCardTypes, plainColumnName } from 'tessa/workflow';
import {
  IKrStageTypeUIHandlerContext,
  KrStageTypeUIHandler,
  StageTypeHandlerDescriptor,
  processManagementDescriptor
} from 'tessa/workflow/krProcess';
import { TypedField, Visibility } from 'tessa/platform';
import { ValidationResult, ValidationResultType } from 'tessa/platform/validation';

import { CardFieldChangedEventArgs } from 'tessa/cards';
import { IControlViewModel } from 'tessa/ui/cards';
import { ProcessManagementStageTypeMode } from '../processManagementStageTypeMode';

/**
 * UI обработчик типа этапа {@link processManagementDescriptor}.
 */
export class ProcessManagementUIHandler extends KrStageTypeUIHandler {
  //#region fields

  private static readonly _modeId: string = plainColumnName(
    'KrProcessManagementStageSettingsVirtual',
    'ModeID'
  );

  private _stageControl?: IControlViewModel;
  private _groupControl?: IControlViewModel;
  private _signalControl?: IControlViewModel;

  //#endregion

  //#region base overrides

  public descriptors(): StageTypeHandlerDescriptor[] {
    return [processManagementDescriptor];
  }

  public async initialize(context: IKrStageTypeUIHandlerContext): Promise<void> {
    const block = context.settingsForms
      .find(i => i.name === DefaultCardTypes.KrProcessManagementStageTypeSettingsTypeName)
      ?.blocks.find(i => i.name === 'MainInfo');

    if (!block) {
      return;
    }

    this._stageControl = block.controls.find(i => i.name === 'StageRow')!;
    this._groupControl = block.controls.find(i => i.name === 'StageGroup')!;
    this._signalControl = block.controls.find(i => i.name === 'Signal')!;

    if (!this._stageControl || !this._groupControl || !this._signalControl) {
      return;
    }

    this.updateVisibility(context.row.tryGetField(ProcessManagementUIHandler._modeId));

    context.row.fieldChanged.add(this.modeChanged);
  }

  public async finalize(context: IKrStageTypeUIHandlerContext): Promise<void> {
    this._stageControl = undefined;
    this._groupControl = undefined;
    this._signalControl = undefined;

    context.row.fieldChanged.remove(this.modeChanged);
  }

  public async validate(context: IKrStageTypeUIHandlerContext): Promise<void> {
    if (context.row.tryGet(ProcessManagementUIHandler._modeId) == undefined) {
      context.validationResult.add(
        ValidationResult.fromText(
          '$KrStages_ProcessManagement_ModeNotSpecified',
          ValidationResultType.Error
        )
      );
    }
  }

  //#endregion

  //#region private methods

  private readonly modeChanged = (args: CardFieldChangedEventArgs): void => {
    if (args.fieldName !== ProcessManagementUIHandler._modeId) {
      return;
    }

    this.updateVisibility(args.fieldTypedValue);
  };

  private updateVisibility(field: TypedField | null | undefined): void {
    this._stageControl!.controlVisibility = Visibility.Collapsed;
    this._groupControl!.controlVisibility = Visibility.Collapsed;
    this._signalControl!.controlVisibility = Visibility.Collapsed;

    if (field) {
      switch (field.$value) {
        case ProcessManagementStageTypeMode.StageMode:
          this._stageControl!.controlVisibility = Visibility.Visible;
          break;
        case ProcessManagementStageTypeMode.GroupMode:
          this._groupControl!.controlVisibility = Visibility.Visible;
          break;
        case ProcessManagementStageTypeMode.SendSignalMode:
          this._signalControl!.controlVisibility = Visibility.Visible;
          break;
      }
    }
  }

  //#endregion
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/StageHandlers/ProcessManagementUIHandler.cs
#nullable enable

using System.Linq;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.UI.Cards;

namespace Tessa.Extensions.Default.Client.UI.KrProcess.StageHandlers
{
    /// <summary>
    /// UI обработчик типа этапа <see cref="StageTypeDescriptors.ProcessManagementDescriptor"/>.
    /// </summary>
    public sealed class ProcessManagementUIHandler :
        StageTypeUIHandlerBase
    {
        #region Fields

        private IControlViewModel? stageControl;
        private IControlViewModel? groupControl;
        private IControlViewModel? signalControl;

        #endregion

        #region Base Overrides

        /// <inheritdoc />
        public override Task Initialize(
            IKrStageTypeUIHandlerContext context)
        {
            // Достаем все необходимые элементы управления
            if (context.SettingsForms.FirstOrDefault(static i => i.Name == DefaultCardTypes.KrProcessManagementStageTypeSettingsTypeName) is not { } form
                || form.Blocks.FirstOrDefault(static i => i.Name == "MainInfo") is not { } block
                || block.Controls.FirstOrDefault(static i => i.Name == "StageRow") is not { } stageControl
                || block.Controls.FirstOrDefault(static i => i.Name == "StageGroup") is not { } groupControl
                || block.Controls.FirstOrDefault(static i => i.Name == "Signal") is not { } signalControl)
            {
                return Task.CompletedTask;
            }

            this.stageControl = stageControl;
            this.groupControl = groupControl;
            this.signalControl = signalControl;

            // Обновляем видимость элементов управления
            this.UpdateVisibility(context.Row.TryGet<int?>(KrConstants.KrProcessManagementStageSettingsVirtual.ModeID));

            // Подписываемся на события изменения режима
            context.Row.FieldChanged += this.ModeChanged;

            return Task.CompletedTask;
        }

        /// <inheritdoc />
        public override Task Finalize(
            IKrStageTypeUIHandlerContext context)
        {
            this.stageControl = null;
            this.groupControl = null;
            this.signalControl = null;

            context.Row.FieldChanged -= this.ModeChanged;

            return Task.CompletedTask;
        }

        /// <inheritdoc />
        public override Task Validate(IKrStageTypeUIHandlerContext context)
        {
            // Проверяем, что режим указан
            if (context.Row.TryGet<int?>(KrConstants.KrProcessManagementStageSettingsVirtual.ModeID) is null)
            {
                // Если режим не указан, то показываем ошибку
                context.ValidationResult.AddError(this, "$KrStages_ProcessManagement_ModeNotSpecified");
            }

            return Task.CompletedTask;
        }

        #endregion

        #region Private Methods

        private void ModeChanged(
            object? _,
            CardFieldChangedEventArgs args)
        {
            if (args.FieldName != KrConstants.KrProcessManagementStageSettingsVirtual.ModeID)
            {
                return;
            }

            this.UpdateVisibility((int?) args.FieldValue);
        }

        private void UpdateVisibility(
            int? modeID)
        {
            this.stageControl!.ControlVisibility = Visibility.Collapsed;
            this.groupControl!.ControlVisibility = Visibility.Collapsed;
            this.signalControl!.ControlVisibility = Visibility.Collapsed;

            switch ((ProcessManagementStageTypeMode?) modeID)
            {
                case ProcessManagementStageTypeMode.StageMode:
                    this.stageControl.ControlVisibility = Visibility.Visible;
                    break;
                case ProcessManagementStageTypeMode.GroupMode:
                    this.groupControl.ControlVisibility = Visibility.Visible;
                    break;
                case ProcessManagementStageTypeMode.SendSignalMode:
                    this.signalControl.ControlVisibility = Visibility.Visible;
                    break;
            }

            // Все контролы расположены в одном блоке.
            this.stageControl.Block.Rearrange();
        }

        #endregion
    }
}
